packages/dx-react-grid-demos/src/demo-sources/grid-row-reordering/tree.jsxt

```diff
@@ -0,0 +1,163 @@
+import React, { useState, useCallback } from 'react';<%&additionalImports%>
+import { TreeDataState, CustomTreeData } from '@devexpress/dx-react-grid';
+import { Plugin, Getter } from '@devexpress/dx-react-core';
+import {
+  Grid,
+  Table,
+  TableHeaderRow,
+  TableTreeColumn,
+} from '@devexpress/dx-react-grid-<%&themeName%>';
+<%&cssImports%>
+
+import {
+  SortableContainer,
+  SortableHandle,
+  SortableElement,
+} from 'react-sortable-hoc';
+import arrayMove from 'array-move';
+
+import {
+  generateRows,
+  defaultColumnValues,
+} from '../../../demo-data/generator';
+
+const getRowId = row => row.id;
+
+const DragHandle = SortableHandle(({ style }) => (
+  <span style={{ ...style, cursor: 'move' }}>{'::::'}</span>
+));
+
+const TableCell = ({ value, ...restProps }) => {
+  const width = restProps.column.name === 'drag' ? 50 : 195;
+  return (
+    <Table.Cell {...restProps} style={{ width }}>
+      {restProps.column.name === 'drag' ? <DragHandle /> : <span>{value}</span>}
+    </Table.Cell>
+  );
+};
+
+const getChildRows = (row, rootRows) => {
+  const childRows = rootRows.filter(r => r.parentId === (row ? row.id : null));
+  return childRows.length ? childRows : null;
+};
+
+const handleOrderedRows = handleRows => ({ rows }) => {
+  setTimeout(() => handleRows(rows));
+  return rows;
+};
+
+const OrderedRowProvider = ({ handleRows }) => (
+  <Plugin
+    name="OrderedRowProvider"
+    dependencies={[{ name: 'CustomTreeData' }]}
+  >
+    <Getter name="rows" computed={handleOrderedRows(handleRows)} />
+  </Plugin>
+);
+
+export default () => {
+  const [columns] = useState([
+    { name: 'drag', title: ' ' },
+    { name: 'name', title: 'Name' },
+    { name: 'gender', title: 'Gender' },
+    { name: 'city', title: 'City' },
+    { name: 'car', title: 'Car' },
+  ]);
+  const [data, setData] = useState(generateRows({
+    columnValues: {
+      id: ({ index }) => index,
+      parentId: ({ index, random }) => (index > 0 ? Math.trunc((random() * index) / 2) : null),
+      ...defaultColumnValues,
+    },
+    length: 20,
+  }));
+  const [tableColumnExtensions] = useState([{ columnName: 'drag', width: 50 }]);
+  const [expandedRowIds, setExpandedRowIds] = useState([0]);
+  const [orderedRows, setOrderedRows] = useState([]);
+
+  const onSortEnd = useCallback(({ newIndex, oldIndex }) => {
+    const rowId = orderedRows[oldIndex].id;
+
+    const source = orderedRows[oldIndex];
+    const sourceIndex = data.indexOf(source);
+    const sourceParentId = source.parentId;
+
+    const target = orderedRows[newIndex];
+    const targetIndex = data.indexOf(target);
+    const targetParentId = target.parentId;
+
+    let newData;
```

Comment:
It is better to move newData calculations outside render method (to leave useCallback here, but to create a separate function which would calculate newData based on `data`).